File: tests/game.js
QUnit.test( "test game blockFitsBounds", function( assert ) {

    const game = new Game();

    assert.ok( game.blockFitsBounds(new Block(0,0,0)) );
    assert.ok( game.blockFitsBounds(new Block(game.dims.x-1,0,0)) );
    assert.ok( !game.blockFitsBounds(new Block(0,game.dims.y,0)) );
});

QUnit.test( "test game translateToBounds", function( assert ) {

    const game = new Game();
    const piece = new Piece([new Block(-2,0,0), new Block(0,-1,0)]);

    assert.ok(!game.pieceFits(piece));

    piece.translate(0,10,0);
    assert.ok(!game.pieceFits(piece));

    game.translateToBounds(piece);
    assert.ok(game.pieceFits(piece));

    const blocks = piece.getBlocks();
    assert.equal(blocks[0].x, 0);
    assert.equal(blocks[0].y, game.dims.y-1);
    assert.equal(blocks[1].x, 2);
    assert.equal(blocks[1].y, game.dims.y-2);
});

QUnit.test( "test game controls", function( assert ) {

    function pieceGenerator() {
        return new Piece([
            new Block(0,0,0),
            new Block(1,0,0),
            new Block(0,1,0),
            new Block(0,0,1)
        ]);
    }
    const game = new Game(pieceGenerator);

    assert.ok(!game.isOver());

    let blocks = game.getActiveBlocks();
    assert.equal(blocks[0].x, Math.round(game.dims.x/2));
    assert.equal(blocks[1].x, Math.round(game.dims.x/2)+1);
    assert.equal(blocks[0].z, game.dims.z-2);
    assert.equal(blocks[3].z, game.dims.z-1);

    assert.ok(game.controls.moveDown());
    assert.ok(game.controls.moveXY(-1, 0));

    blocks = game.getActiveBlocks();
    assert.equal(blocks[0].x, Math.round(game.dims.x/2)-1);
    assert.equal(blocks[1].x, Math.round(game.dims.x/2));
    assert.equal(blocks[0].z, game.dims.z-3);
    assert.equal(blocks[3].z, game.dims.z-2);

    assert.ok(game.controls.rotate('y', -1));
    assert.ok(game.controls.rotate('y', 1));
    assert.ok(game.controls.drop());

    let cem = game.getCementedBlocks();
    assert.equal(cem.length, 4);
    assert.equal(cem[0].x, Math.round(game.dims.x/2)-1);
    assert.equal(cem[0].y, Math.round(game.dims.y/2));
    assert.equal(cem[0].z, 0);
    assert.equal(cem[3].z, 1);

    assert.ok(!game.isOver());

    while(!game.isOver()) {
        game.controls.drop();
    }

    assert.ok(game.isOver());
});

QUnit.test( "test game remove bottom", function( assert ) {

    const dims = new Game().dims;

    function pieceGenerator() {
        // two full layers + one block on top
        let blocks = [new Block(1,1,2)];
        for (let z=0; z<2; ++z) {
            for (let x=0; x<dims.x; ++x) {
                for (let y=0; y<dims.y; ++y) {
                    blocks.push(new Block(x,y,z));
                }
            }
        }

        return new Piece(blocks);
    }

    const game = new Game(pieceGenerator);
    assert.equal(game.getCementedBlocks().length, 0);
    assert.equal(game.getActiveBlocks().length, dims.x*dims.y*2 + 1);

    assert.ok(game.controls.drop());
    assert.ok(!game.isOver());

    let cem = game.getCementedBlocks();
    assert.equal(cem.length, 1);
    assert.equal(cem[0].x, 1);
    assert.equal(cem[0].y, 1);
    assert.equal(cem[0].z, 0);

    assert.ok(game.controls.drop());
    cem = game.getCementedBlocks();
    assert.equal(cem.length, 2 + dims.x*dims.y * 2);

    assert.equal(cem[0].x, 1);
    assert.equal(cem[0].y, 1);
    assert.equal(cem[0].z, 0);
});

File: cpp/src/game.cpp
#include "game.hpp"
#include <cmath>

std::unique_ptr<Game> buildGame(unsigned int randomSeed) {
    return std::unique_ptr<Game>(new ConcreteGame(randomSeed));
}

namespace game_config {
    static const Pos3d DIMENSIONS { 5, 5, 14 };
    static const int DROP_SCORE_MULTIPLIER = 1;
    static const int REMOVAL_SCORE_MULTIPLIER = 20;
    static const int INITIAL_DROP_INTERVAL_MS = 1000;
    static const double DROP_INTERVAL_HALF_LIFE_N_PIECES = 50;
}

ConcreteGame::ConcreteGame(unsigned int randomSeed)
:
    gameBox(game_config::DIMENSIONS),
    blockArray(gameBox),
    pieceGenerator(gameBox, randomSeed),
    activePiece(pieceGenerator.nextPiece()),
    score(0),
    alive(true),
    timeToNextDownMs(game_config::INITIAL_DROP_INTERVAL_MS),
    nDroppedPieces(0)
{}

std::vector<Block> ConcreteGame::getActiveBlocks() const {
    return activePiece.getBlocks();
}

std::vector<Block> ConcreteGame::getCementedBlocks() const {
    return blockArray.getNonEmptyBlocks();
}

std::vector<Block> ConcreteGame::getAllBlocks() const {
    auto blocks = getCementedBlocks();
    auto active = getActiveBlocks();
    blocks.insert(blocks.end(), active.begin(), active.end());
    return blocks;
}

bool ConcreteGame::isOver() const {
    return !alive;
}

int ConcreteGame::getScore() const {
    return score;
}

Pos3d ConcreteGame::getDimensions() const {
    return gameBox.dims;
}

// timed events
bool ConcreteGame::tick(int dtMs) {

    if (isOver()) return false;

    timeToNextDownMs -= dtMs;
    if (timeToNextDownMs <= 0) {
        moveDown();

        // compute next down
        timeToNextDownMs = ceil(
            game_config::INITIAL_DROP_INTERVAL_MS *
            pow(2, -nDroppedPieces /
                game_config::DROP_INTERVAL_HALF_LIFE_N_PIECES));

        return true;
    }
    return false;
}

// controls

bool ConcreteGame::moveXY(int dx, int dy) {
    if (abs(dx) + abs(dy) != 1) abort();
    return setActivePieceIfFits(activePiece.translated(Pos3d {dx,dy,0}));
}

bool ConcreteGame::rotate(Axis axis, RotationDirection dir) {
    if (axis != Axis::X && axis != Axis::Y && axis != Axis::Z) abort();
    if (dir != RotationDirection::CW && dir != RotationDirection::CCW) abort();

    return rotate(Rotation{axis, dir});
}

void ConcreteGame::drop() {
    int height = 0;
    while (moveDown()) height++;
    score += game_config::DROP_SCORE_MULTIPLIER * height;
}

// private helpers
bool ConcreteGame::rotate(Rotation rot) {
    return setActivePieceIfFits(
        gameBox.translateToBounds(activePiece.rotated(rot)));
}

bool ConcreteGame::moveDown() {
    if (!setActivePieceIfFits(activePiece.translated(Pos3d {0,0,-1}))) {
        blockArray.cementPiece(activePiece);
        nDroppedPieces++;

        // remove empty layers
        int nRemoved = 0;
        for (int z = gameBox.dims.z - 1; z >= 0; --z) {
            if (blockArray.isLayerFull(z)) {
                blockArray.removeLayer(z);
                nRemoved++;
            }
        }
        // (2^nRemoved - 1)*C
        // 0 -> 0, 1 -> C, 2 -> 3C, 3 -> 7C, ...
        score += ((1 << nRemoved) - 1) * game_config::REMOVAL_SCORE_MULTIPLIER;

        // new piece, check if fits
        activePiece = pieceGenerator.nextPiece();
        if (!blockArray.pieceFits(activePiece)) {
            alive = false;
        }
        return false;
    }
    return true;
}

bool ConcreteGame::setActivePieceIfFits(const Piece& candidate) {
    if (!blockArray.pieceFits(candidate)) {
        return false;
    }
    activePiece = candidate;
    return true;
}
